packages/account/src/providers/transaction-request/transaction-request.ts

```diff
@@ -694,4 +662,20 @@ export abstract class BaseTransactionRequest implements BaseTransactionRequestLi
       }
     });
   }
+
+  shiftPredicateData() {
```

Comment:
From what I understand, `shiftPredicateData` and `padPredicateData` stem from us setting the predicate data "in stone" _before_ we encode the actual transaction into bytes and this is why both of these functions feel like a workaround. `Predicate.populateTransactionPredicateData` could fall into this category as well. IMO we should definitely refactor the whole transaction encoding workflow so that the dynamic elements of a transaction like `predicateData` get set only when `transaction.encodeTransaction()` is called. This would be another issue, of course; I wanted to get it out here so that it's on our radar.

Reply:
That is not possible right now. 

The `predicateData` needs to be in place when we estimate predicate gas used. It must also be in place when estimating gas and fees before even running the dry-run to get the `gasUsed`.

We need the `predicateData` before even funding the transaction.

Reply:
> That is not possible right now.

Not possible because of the scope of the task or technically not possible? If it's because of the scope of the task, I agree. I don't see technical limitations for it, though.

> The predicateData needs to be in place when we estimate predicate gas used. It must also be in place when estimating gas and fees before even running the dry-run to get the gasUsed.
We need the predicateData before even funding the transaction.

Every transaction has to be converted to bytes at one point before being sent to the node, be it for full execution, for dry-running, or whatever else (e.g. [tx estimation](https://github.com/FuelLabs/fuels-ts/blob/8f62b941193683319181d76a12c7e04effc2e568/packages/account/src/providers/provider.ts#L737)) My comment is  related to this tx-to-bytes conversion: that we resolve all these dynamic aspects of a transaction inside of `TransactionRequest.toTransactionBytes` - at the last moment, just before a transaction is sent.

Reply:
Created #2156 for this where we can continue the discussion.